packages/core/src/Repo.ts

```diff
@@ -0,0 +1,439 @@
+    await this.saveSnapshot(documentId, newDoc)
+
+    // call handlers
+    for (const fn of this.handlers) {
+      await fn(documentId, newDoc)
+    }
+
+    // return the modified document
+    return newDoc
+  }
+
+  /**
+   * Gets the in-memory snapshot of a document
+   * @param documentId
+   * @returns Returns a plain JS object
+   */
+  getSnapshot(documentId: string) {
+    return this.state[documentId]
+  }
+
+  /**
+   * Changes the snapshot of a document synchronously, without modifying the underlying Automerge
+   * changes. This is used to quickly update the UI; the change history can be updated later.
+   * @param documentId
+   * @param fn The change function (usually comes from a ProxyReducer)
+   */
+  changeSnapshot(documentId: string, fn: A.ChangeFn<T>) {
+    // create a new automerge object from the current version's snapshot
+    const oldDoc = this.getSnapshot(documentId) || {}
+    const doc: A.Doc<any> = A.from(oldDoc)
+
+    // apply the change
+    const newDoc = A.change(doc, fn)
+
+    // convert the result back to a plain object
+    const snapshot = { ...newDoc }
+
+    this.setSnapshot(documentId, snapshot)
+    this.log('changed snapshot', documentId, snapshot)
+  }
+
+  /**
+   * Sets the in-memory snapshot of a document. NOTE: This does not update the document's change
+   * history or persist anything; this is just to allow synchronous updates of the state for UI
+   * purposes.
+   * @param documentId
+   * @param snapshot
+   */
+  setSnapshot(documentId: string, snapshot: any) {
+    if (snapshot.DELETED) {
+      this.removeSnapshot(documentId)
+    } else {
+      this.state[documentId] = snapshot
+    }
+  }
+
+  /**
+   * Removes the snapshot with the given `documentId` from in-memory state.
+   * @param documentId
+   */
+  removeSnapshot(documentId: string) {
+    this.log('removeSnapshot', documentId)
+    this.state[documentId] = null
+  }
+
+  /**
+   * Returns the state of the entire repo, containing snapshots of all the documents.
+   */
+  getState(): RepoSnapshot<T> {
+    return this.state
+  }
+
+  /**
+   * Replaces the (snapshot) state of the entire repo. NOTE: This doesn't update the repo's change
+   * history or persist anything; this is only used for synchronous updates of the state for UI
+   * purposes.
+   */
+  loadState(replacementState: RepoSnapshot<T>) {
+    this.state = replacementState
+  }
+
+  /**
+   * Used for sending the entire current state of the repo to a new peer.
+   * @returns Returns an object mapping documentIds to an array of changes.
+   */
+  async getHistory(): Promise<RepoHistory> {
+    const history: RepoHistory = {}
+
+    const index = this.database!.transaction('feeds').store.index('documentId')
+    const changeSets = index.iterate(undefined, 'next')
+
+    for await (const cursor of changeSets) {
+      const { documentId, changes } = cursor.value as ChangeSet
+      history[documentId] = (history[documentId] || []).concat(changes)
```

Comment:
Hopefully the other PR will take care of this, but loading everything in memory defeats the purpose of the cursor.

Reply:
The solution for large datasets is to send history in batches of N docs, with N to be determined experimentally. I've added TODOs. 
189d8bea9bf69187d0188307611f14cefbc77169